Adds demo on unitary designs

glassnotes opened the pull request:
Companion demo to the Haar measure one.

TODO:
 - [x] formal progression from spherical to unitary designs
 - [x] description of Pauli/Clifford group
 - [x] finalize computational experiment examples
 - [x] rewrite experiment using QNode transforms (need to change to transform decorator branch for it to build)
 - [x] conclusions
 - [x] references

glassnotes commented on requirements.txt:
```diff
@@ -10,7 +10,7 @@ sympy==1.6.2
 pip==20.1
 pyscf==1.7.2
-git+https://github.com/PennyLaneAI/pennylane.git@master
+git+https://github.com/PennyLaneAI/pennylane.git@better-transforms
```
Note to self: revert this change once the transforms branch is merged in.

glassnotes replied:
```suggestion
git+https://github.com/PennyLaneAI/pennylane.git@master
```

josh146 commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,558 @@
+# we create a list of tapes, one for each Clifford. We then execute all
+# the tapes on the device, and compute the fidelity.
+
+# Note: This whole section is a good place for a QNode transform since
+# it's a list of tapes, and we are doing something with the execution results.
```
Nice! Something like this?

```python

@qml.single_tape_transform
def add_noisy_operation_sequence(tape, damp_factor, depo_factor, flip_prob):
    """Add the noisy operation sequence to a tape prior to measurement"""
    for op in tape.operations:
        op.queue()

    noisy_operation_sequence(damp_factor, depo_factor, flip_prob)

    for m in tape.measurements:
        m.queue()


@qml.single_tape_transform
def return_state(tape):
    """Modify a tape to simply return the quantum state"""
    for op in tape.operations:
        op.queue()

    qml.state()


@qml.qnode_transform
def fidelity_comparison(qnode, damp_factor, depo_factor, flip_prob, single_qubit_cliffords):
    tape = return_state(qnode.qtape)

    ideal_tapes = []
    noisy_tapes = []

    for clifford in single_qubit_cliffords:
        # Set up the ideal tape
        twirled_tape = conjugate_with_clifford(tape, clifford)
        ideal_tapes.append(twirled_tape)

        # Set up the noisy tape
        noisy_tape = add_noisy_operation_sequence(tape, damp_factor, depo_factor, flip_prob)
        twirled_tape = conjugate_with_clifford(noisy_tape, clifford)
        noisy_tapes.append(twirled_tape)

    def fidelity_fn(res):
        # Use the results to compute the fidelities
        fidelities = []
        ideal_results = res[:len(ideal_tapes)]
        noisy_results = res[len(ideal_tapes):]

        for state_1, state_2 in zip(ideal_results, noisy_results):
            fidelities.append(fidelity(state_1[0], state_2[0]))

        return fidelities

    return ideal_tapes + noisy_tapes, fidelity_fn


@qml.qnode(dev)
def test_qnode():
    qml.SX(wires=0)
    return qml.expval(qml.PauliZ(0))

fidelity = fidelity_comparison(test_qnode)
fidelity(damp_factor, depo_factor, flip_prob, single_qubit_cliffords)
```

glassnotes replied:
Yes this is similar to what I was thinking. I played around with it a bit more and, given that we haven't quite ironed out everything with QNode transforms, I've rewritten it to use qfunc transforms instead. The result is really nice because it's a lot more modular, and for the most part eliminates the need for tapes except for within the quantum function transforms explicitly.

josh146 replied:
Oh nice! I just had a look, I was super curious. It seems much more concise!

One question though: how come you made `apply_single_clifford(tape, clifford_string, inverse=False)` a tape transform, could it just be a simple qfunc? e.g., remove the decorator and the unused `tape` argument?

```python
def apply_single_clifford(clifford_string, inverse=False):
    for gate in clifford_string:
        if gate == 'H':
            qml.Hadamard(wires=0)
        else:
            if inverse:
                qml.PhaseShift(np.pi/2, wires=0)
            else:
                qml.PhaseShift(-np.pi/2, wires=0)
```

josh146 replied:
I just deleted my comment because I was looking at the old build version, not the new one 🙈

glassnotes commented on demonstrations/tutorial_haar_measure.py:
```diff
@@ -11,11 +11,13 @@
     tutorial_barren_plateaus Barren plateaus
 
 
+
```
```suggestion

```

glassnotes commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+
+    tutorial_haar_measure Understanding the Haar measure
+
+*Author: PennyLane dev team. Posted: XX May 2021. Last updated: XX May 2021.*
```
Note to self: update the date before merging.

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+
+"""
+
+single_qubit_cliffords = [
```
Wait, I'm not sure I understand this. How exactly do the ways that X and Z can be transformed lead to these products of S and H defining the Clifford group?

glassnotes replied:
These are just the 24 unique elements generated by H and S. We can see why there are 24 using the combinatorial argument above, but then it takes some work to generate the set. There are two approaches, one of which is to look at all possible ways that X and Z can be mapped to {+/-X, +/- Y, +/-Z}, and determine which Clifford performs each mapping. The other is simpler and is to just multiply together generators until you get 24 unique combinations.  In any case I'll work on improving this explanation :+1:

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+]
+
+######################################################################
+# .. admonition:: Fun fact
```
This text is great, but I would remove the "fun fact" format and leave it as a normal paragraph

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+np.random.seed(42)
+
+# Use the mixed state simulator
+dev = qml.device("default.mixed", wires=1)
```
:star_struck: I'm always happy to see defaut.mixed being used :grin:

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+depo_factor = 0.02
+flip_prob = 0.01
+
+noisy_experiment = apply_noise(damp_factor, depo_factor, flip_prob)(ideal_experiment)
```
This is so beautiful! :heart_eyes:

glassnotes replied:
I know :heart_eyes: Being able to use the transforms like this has streamlined the code so much, huge improvement over the original version.

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+        if gate == 'H':
+            qml.Hadamard(wires=0)
+        else:
+            # We must manually apply the inverse as it is not yet supported on default.mixed
```
:crying_cat_face:

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+# *Approximate* unitary :math:`t`-designs (where the equality in the definition
+# is replaced by approximately equal up to some finite precision) are also of
+# interest, as there ways to construct them that are more efficient than those
+# of exact designs [#Dankert]_. It particular, it has been shown that
```
```suggestion
# of exact designs [#Dankert]_. In particular, it has been shown that
```

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+# interest, as there ways to construct them that are more efficient than those
+# of exact designs [#Dankert]_. It particular, it has been shown that
+# approximate complex projective 4-designs have applications to the state
+# distinction problem [#Ambainis]_.
```
State distinction instead of state discrimination? First time I hear this. Is it because discrimination is a loaded word?

glassnotes replied:
Ah it's just called "state distinction" in that reference and a couple others that refer to it as an application of 4-designs; I thought maybe it was a slightly different task than state discrimination? I will look into it :+1:

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+# have a direct correspondence with mutually unbiased bases, which are optimal
+# quantum measurements [#Gaeta]_ (as well as complex projective designs); and Latin
+# squares themselves have direct correspondence with affine and projective
+# planes, bringing us full circle back to the Fano plane from which we began.
```
Full triangle :wink:

glassnotes replied:
Love it :smiling_imp:

ixfoduap reviewed (changes requested):
Another outstanding tutorial, wow! :star_struck: 

The tutorial could go live as is after correcting a couple of typos. However, I am hoping it can get even better. My main comments are:

- I would like to see a short explanation/example of why designs work. How come we can get an exact answer by evaluating on only a few points?
-  I felt that the flow was broken in different parts, specially after a really, really great introduction. I think the tutorial may be trying to do too much (overly enthusiastic) and could benefit from a bit more focus and compactness
- The images in the beginning are so great! I wish there were more like this throughout

glassnotes commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+Spherical designs exist for all :math:`t` and dimension :math:`d`
+[#Handbook]_. They are not always unique, and may have varying numbers of points
+(as can be seen above). Furthermore, a spherical :math:`t`-design is a
+:math:`k`-design for all :math:`k < t` as well!
```
I wasn't kidding about the overly-enthusiastic part! :sweat_smile:  But yes good point, I will tone it down :grin:

glassnotes commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+(as can be seen above). Furthermore, a spherical :math:`t`-design is a
+:math:`k`-design for all :math:`k < t` as well!
+
+
```
That's a good point. I'll work on putting an explicit example of a calculation with a spherical design, I think that will be really helpful for illustrating the point (and also help break up the long amount of text that comes before the main code example).

glassnotes commented:
> Another outstanding tutorial, wow! star_struck
> 
> The tutorial could go live as is after correcting a couple of typos. However, I am hoping it can get even better. My main comments are:
> 
>     * I would like to see a short explanation/example of why designs work. How come we can get an exact answer by evaluating on only a few points?
> 
>     * I felt that the flow was broken in different parts, specially after a really, really great introduction. I think the tutorial may be trying to do too much (overly enthusiastic) and could benefit from a bit more focus and compactness
> 
>     * The images in the beginning are so great! I wish there were more like this throughout

@ixfoduap ready for re-review. I've made the following major changes to address your comments:
 - added an example for the spherical designs
 - removed the section about complex projective designs (helps the demo get to the point faster, plus I wasn't actually doing anything with them; I've left the idea in one of the callout boxes)
 - added a bit more intuition behind how the Clifford group works and is constructed

In terms of the definitions, it feels too hand-wavy if I just omit them; it's important to show that `=` sign between the finite sum and the integral. But now that I've removed complex projective design, each definition is preceded by some motivation and associated with an explicit example, which hopefully provides enough detail to grasp the concept even if some of the formal math is fuzzy.

glassnotes commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,692 @@
+
+[TODO: picture will be updated with some additional solids]
+
+We see from these illustrations that spherical designs are sets of
```
Haven't said exactly this, but tried to rework the contents.

glassnotes commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,746 @@
+# This clearly differs from the true value. We need a design with :math:`t=4`
+# or better in order to compute this average, and when such a design is
+# available, we may save significant computational time.
+# 
```
This could be a nice spot for another graphic that shows a cube and a dodecahedron; the earlier graphic can be modified so that they are swapped out for an octahedron and icosahedron respectively. Allows us to show more examples of shapes without cluttering the original drawing.

glassnotes commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,746 @@
+print(f"Mean fidelity      = {fid_mean}")
+print(f"Std. dev. fidelity = {fid_std}")
+
+plt.hist(fidelities, bins=20)
```
I'm considering to remove this histogram - it's neat to see the distribution, but I don't actually use any results about the distribution save for the average.

glassnotes commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,746 @@
+# interest, as there ways to construct them that are more efficient than those
+# of exact designs [#Dankert]_. In particular, it has been shown that
+# approximate complex projective 4-designs have applications to the state
+# distinction problem [#Ambainis]_.
```
```suggestion
# discrimination problem [#Ambainis]_.
```

glassnotes commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -0,0 +1,746 @@
+
+    tutorial_haar_measure Understanding the Haar measure
+
+*Author: PennyLane dev team. Posted: XX May 2021. Last updated: XX May 2021.*
```
```suggestion
*Author: PennyLane dev team. Posted: XX June 2021. Last updated: XX June 2021.*
```

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -70,20 +70,23 @@
+Before diving into unitary designs, let's look at the sphere for some
+intuition.  Suppose we have a polynomial in :math:`d` variables, and we would
+like to compute its average over the surface of a real, :math:`d`-dimensional
+unit sphere, :math:`S(R^d)`. One can do so by integrating that function over the
```
Consider using "we" instead of "one" for a tutorial.

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -70,20 +70,23 @@
+sphere (using the proper measure), but that would be a lot of parameters to
+keep track of. 
+
+One could alternatively approximate the average by sampling thousands of points
```
Great paragraph! For me it makes the connection with Monte Carlo integration very clear.

In fact, it makes me think that t-designs are a form of quasi-Monte Carlo, which involve strategies that perform the integration over a fixed set of points, e.g., evenly spaced. The magic here is that they can be chosen so the result is exact

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -122,223 +128,266 @@
+quickly due to a higher degree. In 3 dimensions, a function with degree 2 can
+be approximated with something as simple as a tetrahedron, despite the fact that
+it doesn't look spherical at all, while more complex functions require more
+points and thus more complex configurations for the design.  Spherical designs
```
```suggestion
points and thus more intricate configurations for the design.  Spherical designs
```
or other synonym to avoid repeated use of complex

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -122,223 +128,266 @@
-:math:`n`-qubit Clifford group, :math:`\mathcal{C}(n)`, is the *normalizer* of the
-Pauli group. In simpler terms, the Clifford group is the set of operations that
-send Paulis to Paulis when acting under conjugation (up to a phase), i.e.,
+# The golden ratio
```
:star_struck:

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -122,223 +128,266 @@
-generated by :math:`H` and :math:`S`).  For example, consider the action of
-:math:`HS`:
+######################################################################
+# This is exactly the value we expect. What happens if we try to do this using
```
LOOOOOVE this example!

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -122,223 +128,266 @@
+# Unitary designs
+# ^^^^^^^^^^^^^^^
+# 
+# We've demonstrated how spherical are sets of evenly-spaced points on a sphere
```
Revise this sentence

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -122,223 +128,266 @@
+#    S X S^\dagger = Y, \quad S Y S^\dagger = -X, \quad S Z S^\dagger = Z.
+#
+# If both :math:`H` and :math:`S` map Paulis to Paulis, then products of them
+# do so as well (in group theory terms, the single-qubit Clifford group is
```
Consider putting the sentence in parenthesis as a regular new sentence, without parenthesis

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -351,31 +400,36 @@
+# is balanced in the sense that each eigenstate is obtained the same number of
+# times.
+#
+# While we do not explore it here, the multi-qubit Clifford can also be
```
```suggestion
# The multi-qubit Clifford can also be
```

ixfoduap reviewed (changes requested):
I really love the changes!

My only remaining suggestions:
- The tutorial could use more images like the ones appearing in the beginning. Maybe you're working on them already?
- I think you commented that it may be a good idea to remove the plot with the distribution of averages. I agree, I think it's best to just quote the average and standard deviation

glassnotes commented:
> I really love the changes!
> 
> My only remaining suggestions:
> 
>     * The tutorial could use more images like the ones appearing in the beginning. Maybe you're working on them already?
> 
>     * I think you commented that it may be a good idea to remove the plot with the distribution of averages. I agree, I think it's best to just quote the average and standard deviation

Done, and done! Added two new colourful graphics. Let me know if you think of any other places where one could be nice.

ixfoduap commented on demonstrations/tutorial_unitary_designs.py:
```diff
@@ -216,14 +214,14 @@ def f(x, y, z):
-# polynomials that are functions of the entries of unitary matrices [#Dankert]_,
-# [#Gross]_.
+# We've learned now that spherical designs are sets of evenly-spaced points, and
+# saw how they can be used as a short cut to evaluate the average of a
```
```suggestion
# saw how they can be used as a shortcut to evaluate the average of a
```
I think short cut is used too, but I've always thought of it as a single word

glassnotes replied:
Oh you're right! :exploding_head:

ixfoduap reviewed (approved):
:shipit:
